src/reconcile/rec_visibility.c

```diff
@@ -357,6 +398,58 @@ __wt_rec_upd_select(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_INSERT *ins, v
             break;
     }
 
+    /*
+     * Check whether the earliest update in the chain is out-of-order in relation to the current
+     * on-disk value. If the chain is out-of-order, the updates need to be pinned to the cache.
+     *
+     * If the previous update exists on the disk in some form (either because the update has been
+     * restored from the disk OR it has been marked as having already been written to the history
+     * store), then we know that the update has already been made durable so we shouldn't worry
+     * about trying to pin it to the cache.
+     */
+    if (!F_ISSET(r, WT_REC_IN_MEMORY) && prev_upd != NULL && prev_upd->txnid != WT_TXN_ABORTED &&
+      !F_ISSET(prev_upd, WT_UPDATE_HS | WT_UPDATE_RESTORED_FROM_DS | WT_UPDATE_RESTORED_FROM_HS) &&
+      vpack != NULL) {
+        /*
+         * If the on-disk value has a stop time point, then let's compare with the implicit
+         * tombstone that it represents.
+         */
+        if (WT_TIME_WINDOW_HAS_STOP(&vpack->tw)) {
+            cmp_ts = vpack->tw.stop_ts;
+            cmp_txnid = vpack->tw.stop_txn;
+
+            /*
+             * If we had pinned updates last reconciliation, it's possible that the tombstone and
+             * update that form the onpage value may still be in the update list. If that's the
+             * case, we can't expect the last update (which corresponds to the start time point of
+             * the onpage value), to be later than the stop time point. We can easily check if that
+             * is the case by comparing the last update in the chain with the start time point.
+             */
+            if (prev_upd->start_ts == vpack->tw.start_ts &&
+              prev_upd->txnid == vpack->tw.start_txn) {
+                cmp_ts = WT_TS_NONE;
+                cmp_txnid = WT_TXN_NONE;
+            }
+        } else {
+            cmp_ts = vpack->tw.start_ts;
+            cmp_txnid = vpack->tw.start_txn;
+        }
+        if ((WT_TXNID_LT(prev_upd->txnid, cmp_txnid) ||
+              (prev_upd->start_ts != WT_TS_NONE && prev_upd->start_ts < cmp_ts)) &&
+          !__wt_txn_upd_visible_all(session, prev_upd)) {
+            upd_select->upd = NULL;
+            has_pinned_updates = true;
+
+            /*
+             * Again, if we're pinning this update to the cache instead of writing it to the disk
+             * because it is out-of-order, then it cannot be stable. If our checkpoint is skipping
+             * writing stable updates to the disk then we need to know about it.
+             */
+            WT_ASSERT(session, !S2C(session)->txn_global.has_stable_timestamp ||
+                S2C(session)->txn_global.stable_timestamp > prev_upd->start_ts);
```

Comment:
I don't understand this. If we want to ensure checkpoint not skipping stable updates, should the condition be S2C(session)->txn_global.stable_timestamp < prev_upd->start_ts?

Reply:
be51f88